src/sys/unix/selector/poll.rs

```diff
@@ -0,0 +1,725 @@
+    poll_fds_index: usize,
+    /// The key of the `Event` associated with this file descriptor.
+    token: Token,
+    /// Used to communicate with IoSourceState when we need to internally deregister
+    /// based on a closed fd.
+    shared_record: Arc<RegistrationRecord>,
+}
+
+impl SelectorState {
+    pub fn new() -> io::Result<SelectorState> {
+        let notify_waker = WakerInternal::new()?;
+
+        Ok(Self {
+            fds: Mutex::new(Fds {
+                poll_fds: vec![PollFd(libc::pollfd {
+                    fd: notify_waker.as_raw_fd(),
+                    events: libc::POLLIN,
+                    revents: 0,
+                })],
+                fd_data: HashMap::new(),
+            }),
+            pending_removal: Mutex::new(Vec::new()),
+            pending_wake_tokens: Mutex::new(Vec::new()),
+            notify_waker,
+            waiting_operations: AtomicUsize::new(0),
+            operations_complete: Condvar::new(),
+            #[cfg(debug_assertions)]
+            id: NEXT_ID.fetch_add(1, Ordering::Relaxed),
+        })
+    }
+
+    pub fn select(&self, events: &mut Events, timeout: Option<Duration>) -> io::Result<()> {
+        let deadline = timeout.map(|t| Instant::now() + t);
+
+        events.clear();
+
+        let mut fds = self.fds.lock().unwrap();
+        let mut closed_raw_fds = Vec::new();
+
+        loop {
+            // Complete all current operations.
+            loop {
+                if self.waiting_operations.load(Ordering::SeqCst) == 0 {
+                    break;
+                }
+
+                fds = self.operations_complete.wait(fds).unwrap();
+            }
+
+            // Perform the poll.
+            trace!("Polling on {:?}", &fds);
+            let num_events = poll(&mut fds.poll_fds, deadline)?;
+            if num_events == 0 && deadline.map(|v| v <= Instant::now()).unwrap_or(false) {
+                // timeout
+                return Ok(());
+            }
+
+            trace!("Poll finished: {:?}", &fds);
+            let notified_events = fds.poll_fds[0].0.revents;
+            let notified = notified_events != 0;
+            let mut num_fd_events = if notified { num_events - 1 } else { num_events };
+
+            let mut pending_wake_tokens_guard = self.pending_wake_tokens.lock().unwrap();
+            let pending_wake_tokens =
+                std::mem::replace(pending_wake_tokens_guard.as_mut(), Vec::new());
+            drop(pending_wake_tokens_guard);
+
+            if notified {
+                self.notify_waker.ack_and_reset();
+                num_fd_events += pending_wake_tokens.len();
+            }
+
+            // We now check whether this poll was performed with descriptors which were pending
+            // for removal and filter out any matching.
+            let mut pending_removal_guard = self.pending_removal.lock().unwrap();
+            let mut pending_removal = std::mem::replace(pending_removal_guard.as_mut(), Vec::new());
```

Comment:
Same as above, I'm worried about how prone this code is to dead locks. Ofc it's possible to fix, just hard to create robust tests for the edge cases

Reply:
I'm not saying we add any locks or keep the lock for longer. I'm say instead of using `Vec::new` we add that vector to `Selector` (not `SelectorState`) and do `swap(pending_removal_guard, self.pending_removal_buffer)`, this way we reuse the allocation and we can keep processing it without holding the lock.

Reply:
Also see my comment about `Selector::selector` taking a mutable self reference.